src/nnet3/discriminative-supervision.cc

```diff
@@ -0,0 +1,528 @@
+  // ensure this is equal to the acoustic scale used while training. This is 
+  // because, on splitting lattices, the initial and final costs are added 
+  // into the graph cost.
+  KALDI_ASSERT(config_.supervision_config.acoustic_scale != 0.0);
+  if (config_.supervision_config.acoustic_scale != 1.0)
+    fst::ScaleLattice(fst::AcousticLatticeScale(
+          config_.supervision_config.acoustic_scale), lat);
+
+  LatticeStateTimes(*lat, &(scores->state_times));
+  int32 num_states = lat->NumStates();
+  std::vector<std::pair<int32,int32> > state_time_indexes(num_states);
+  for (int32 s = 0; s < num_states; s++) {
+    state_time_indexes[s] = std::make_pair(scores->state_times[s], s);
+  }
+
+  // Order the states based on the state times. This is stronger than just
+  // topological sort. This is required by the lattice splitting code.
+  std::sort(state_time_indexes.begin(), state_time_indexes.end());
+  
+  std::vector<int32> state_order(num_states);
+  for (int32 s = 0; s < num_states; s++) {
+    state_order[state_time_indexes[s].second] = s;
+  }
+
+  fst::StateSort(lat, state_order);
+  ComputeLatticeScores(*lat, scores);
+}
+
+void DiscriminativeSupervisionSplitter::ComputeLatticeScores(const Lattice &lat,
+    LatticeInfo *scores) const {
+  LatticeStateTimes(lat, &(scores->state_times));
+  ComputeLatticeAlphasAndBetas(lat, false, 
+                               &(scores->alpha_p), &(scores->beta_p));
+  scores->Check();  
+  // This check will fail if the lattice is not breadth-first search sorted
+}
+
+void AppendSupervision(const std::vector<const DiscriminativeSupervision*> &input,
+    bool compactify,
+    std::vector<DiscriminativeSupervision> *output_supervision) {
+  KALDI_ASSERT(!input.empty());
+  int32 num_inputs = input.size();
+  if (num_inputs == 1) {
+    output_supervision->resize(1);
+    (*output_supervision)[0] = *(input[0]);
+    return;
+  }
+  output_supervision->clear();
+  output_supervision->reserve(input.size());
+  for (int32 i = 0; i < input.size(); i++) {
+    const DiscriminativeSupervision &src = *(input[i]);
+    KALDI_ASSERT(src.num_sequences == 1);
+    if (compactify && !output_supervision->empty() &&
+        output_supervision->back().weight == src.weight &&
+        output_supervision->back().frames_per_sequence ==
+        src.frames_per_sequence) {
+      // Combine with current output
+      // append src.den_lat to output_supervision->den_lat.
+      AppendLattice(&output_supervision->back().den_lat, src.den_lat);
+
+      output_supervision->back().num_ali.insert(
+          output_supervision->back().num_ali.end(), 
+          src.num_ali.begin(), src.num_ali.end());
+
+      output_supervision->back().num_sequences++;
+    } else {
+      output_supervision->resize(output_supervision->size() + 1);
+      output_supervision->back() = src;
+    }
+
+    output_supervision->back().Check();
+  }
+}
+
+void AppendLattice(Lattice *lat, const Lattice &src_lat) {
```

Comment:
there is a lot of lava-flow here.
state times are not needed, and your comment is not needed.
But why use this function at all?  Just call Concat and then TopSort.